### PR TITLE
Adds Block.prototype.mixin() and Blockly.Extensions.registerMixin().

### PR DESCRIPTION
**Pull request by AnmAtAnm**
This adds support for a common use pattern in extensions, and adds error checking to avoid future incompatibilities.

Includes tests.

### CODE REVIEW

File: core/block.js [hunk 1]
```diff
@@ -1043,6 +1043,33 @@ Blockly.Block.prototype.jsonInit = function(json) {
...
+  }
+  if (!optDisableCheck) {
+    var overwrites = [];
+    for (var key in mixinObj) {
+      if (this[key] !== undefined) {
+        overwrites.push(key);
+      }
+    }
+    if (overwrites.length > 0) {
+      throw new Error("Mixin will overwrite block members: " + overwrites.toString());
```
Comment (NeilFraser) [inline core/block.js]: No need for '> 0' (that's speaking JS with a Java accent).
Also no need for '.toString()'.
Single quotes prefered.
Style guide states: Optional function arguments start with opt_.

File: core/extensions.js [hunk 2]
```diff
@@ -64,6 +64,18 @@ Blockly.Extensions.register = function(name, initFn) {
...
+ * @param {string} name The name of this extension.
+ * @param {!Object} mixinObj The values to mix in.
+ * @throws {Error} if the extension name is empty, the extension is already
+ *     registered.
+ */
+Blockly.Extensions.registerMixin = function(name, mixinObj) {
+  Blockly.Extensions.register(name, function() {
+    this.mixin(mixinObj);
+  });
+};
```
Comment (rachel-fenichel) [inline core/extensions.js]: nit: missing newline

File: core/extensions.js [hunk 3]
```diff
@@ -64,6 +64,18 @@ Blockly.Extensions.register = function(name, initFn) {
 };
 
 /**
+ * Registers a new extension function that adds all key/value of mixinObj.
+ * @param {string} name The name of this extension.
+ * @param {!Object} mixinObj The values to mix in.
+ * @throws {Error} if the extension name is empty, the extension is already
```
Comment (rachel-fenichel) [inline core/extensions.js]: nit: did you mean "if the extension name is empty *or* the extension is already registered"?

File: core/extensions.js (hunk 2 above)
Comment (AnmAtAnm) [inline core/extensions.js]: Fixed.

File: core/extensions.js (hunk 3 above)
Comment (AnmAtAnm) [inline core/extensions.js]: Yes. Fixed.

File: core/extensions.js (hunk 2 above)
Comment (AnmAtAnm) [inline core/extensions.js]: Fixed.

**Pull request merged**